patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs

```diff
@@ -0,0 +1,182 @@
+		}*/
+
+
+		/// <summary>
+		/// Adds the specified ingredients to this recipe with the given item types.
+		/// Ex.: 
+		/// <example>recipe.AddIngredient(ItemID.IronAxe, ItemID.IronPickaxe)</example>
+		/// </summary>
+		/// <param name="types">The item types.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(params int[] types)
+		{
+			for (int i = 0; i < types.Length; i++)
+				Requires(types[i]);
+
+
+			return this;
+		}
+
+
+		/// <summary>Adds an ingredient to this recipe with the given item name from the given mod, and with the given stack stack. If the mod parameter is null, then it will automatically use an item from the mod creating this recipe.</summary>
+		/// <param name="mod">The mod.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">The item " + itemName + " does not exist in mod " + mod.Name + ". If you are trying to use a vanilla item, try removing the first argument.</exception>
+		public RecipeBuilder Requires(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe.AddIngredient(mod, itemName, stack);
+
+			return this;
+		}
+
+		/// <summary>Adds an ingredient to this recipe of the given type and stack size.</summary>
+		/// <typeparam name="T">The type.</typeparam>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires<T>(int stack = 1) where T : ModItem => Requires(ModContent.ItemType<T>(), stack);
+
+
+		/// <summary>Adds an ingredient to this recipe of the given type of item and stack size.</summary>
+		/// <param name="modItem">The item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(ModItem modItem, int stack = 1)
+		{
+			_recipe.AddIngredient(modItem, stack);
+
+			return this;
+		}
+
+
+		/// <summary>Adds a recipe group ingredient to this recipe with the given RecipeGroup name and stack size. Vanilla recipe groups consist of "Wood", "IronBar", "PresurePlate", "Sand", and "Fragment".</summary>
+		/// <param name="recipeGroup">The name.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException"></exception>
+		public RecipeBuilder Requires(string recipeGroup, int stack = 1)
+		{
+			_recipe.AddRecipeGroup(recipeGroup, stack);
+
+			return this;
+		}
+
+
+		/// <summary>
+		/// Adds one or many required crafting station(s) with the given tile type(s) to the recipe being built.
+		/// Ex.:
+		/// <example>At(TileID.WorkBenches, TileID.Anvils)</example>
+		/// </summary>
+		/// <param name="tileTypes"></param>
+		/// <returns></returns>
+		public RecipeBuilder At(params int[] tileTypes)
+		{
+			for (int i = 0; i < tileTypes.Length; i++)
+				_recipe.AddTile(tileTypes[i]);
+
+			return this;
+		}
+
+
+		public RecipeBuilder RequiresLava() => NeedVar(recipe => recipe.needLava = true);
```

Comment:
no need for an Action and a common method here. Yes it lets you use expression methods. No it makes it harder to read.
```
public RecipeBuilder RequiresLava() {
    _recipe.needLave = true;
    return this;
}
```